src/mono/mono/mini/aot-compiler.c

```diff
@@ -14355,6 +14382,50 @@ add_interp_in_wrapper_for_sig (MonoAotCompile *acfg, MonoMethodSignature *sig)
 	add_method (acfg, wrapper);
 }
 
+static gboolean
+mono_aot_add_direct_pinvoke (MonoAotCompile *acfg, char *dpi)
+{
+	char *direct_pinvoke = g_strdup (dpi);
+	if (!direct_pinvoke)
+		return TRUE;
+
+	if (direct_pinvoke[0] == '#')
+		return TRUE;
+
+	g_strstrip (direct_pinvoke);
+	if (!direct_pinvoke)
+		return TRUE;
+
+	char **direct_pinvoke_split = g_strsplit (direct_pinvoke, "!", -1);
+	char *library_name = g_strdup (direct_pinvoke_split[0]);
+	if (!direct_pinvoke_split[1]) {
+		g_hash_table_insert (acfg->direct_pinvokes, library_name, NULL);
+		g_strfreev (direct_pinvoke_split);
+		return TRUE;
+	}
+	if (direct_pinvoke_split[2]) {
+		aot_printerrf (acfg, "The provided 'direct_pinvoke' AOT option '%s' shouldn't contain multiple '!'\n", dpi);
+		g_strfreev (direct_pinvoke_split);
+		return FALSE;
+	}
+
+	char *entrypoint_name = g_strdup (direct_pinvoke_split[1]);
+	if (g_hash_table_contains (acfg->direct_pinvokes, library_name)) {
+		GHashTable *val = g_hash_table_lookup (acfg->direct_pinvokes, library_name);
+		if (!val || g_hash_table_contains (val, entrypoint_name)) {
+			g_strfreev (direct_pinvoke_split);
+			return TRUE;
+		}
+		g_hash_table_insert (val, entrypoint_name, NULL);
+	} else {
+		GHashTable *val = g_hash_table_new_full (NULL, g_str_equal, g_free, NULL);
+		g_hash_table_insert (val, entrypoint_name, NULL);
+		g_hash_table_insert (acfg->direct_pinvokes, library_name, val);
+	}
+	g_strfreev (direct_pinvoke_split);
+	return TRUE;
```

Comment:
Lots of stuff can leak in this function due to multiple exit points and lack of memory ownership control. Maybe you should have an exit label instead where you clean up and when you transfer ownership you set the variable to NULL.

Reply:
Added two labels
```
cleanup:
	g_strfreev (direct_pinvoke_split);

early_exit:
	return processed;
```
where processed is whether or not processing the specified direct pinvoke was successful.

It was [recommended to `g_strfreev` the newly allocated array of strings](https://docs.gtk.org/glib/func.strsplit.html). Would it be preferrable to instead have `direct_pinvoke_split = NULL`?

Reply:
Note sure I understand what you are asking, you should free all allocated memory that has not moved ownership into hash table before exiting the function.

Reply:
Ah right, I was looking to free the allocated memory for this `char **direct_pinvoke_split` by using `g_strfreev (direct_pinvoke_split)`. I thought you were saying to use
```
direct_pinvoke_split = NULL
return processed;
```
to free the allocated memory instead.

Reply:
So you have different kind of allocated memory in this function, memory allocated by g_strsplit should be deallocated with g_strfreev, memory allocated with g_strdup should be deallocated with g_free. Memory that moved into hash table will be deallocated when hash table gets freed, so you need to handle memory ownership as well.

Reply:
I think I have a slightly better understanding now.
After refactoring the umbrella function that processes specific direct pinvokes, 
whenever a `char *` is allocated, I `g_free` it and set the pointer to NULL (is setting the pointer to NULL redundant if its a local var?)
whenever the `char **` is allocated from `g_strsplit`, I `g_strfreev` it and set the `char **` pointer to null (I'm thinking that `g_strfreev` will set each individual `char *` pointer to `NULL`)
when passing the `char *` memory ownership to the final `add_direct_pinvoke` method (should be the only time when the module and entrypoint pointers are not `NULL`), afterwards, I free the strings and set the `char **` and `char *` to NULL.